ext/opentelemetry-ext-cloud-trace/src/opentelemetry/ext/cloud_trace/__init__.py

```diff
@@ -0,0 +1,264 @@
+        self, spans: Sequence[Span]
+    ) -> List[Dict[str, Any]]:
+        """Translate the spans to Cloud Trace format.
+
+        Args:
+            spans: Tuple of spans to convert
+        """
+
+        cloud_trace_spans = []
+
+        for span in spans:
+            ctx = span.get_context()
+            trace_id = _get_hexidecimal_trace_id(ctx.trace_id)
+            span_id = _get_hexidecimal_span_id(ctx.span_id)
+            span_name = "projects/{}/traces/{}/spans/{}".format(
+                self.project_id, trace_id, span_id
+            )
+
+            parent_id = None
+            if span.parent:
+                parent_id = _get_hexidecimal_span_id(span.parent.span_id)
+
+            start_time = _get_time_from_ns(span.start_time)
+            end_time = _get_time_from_ns(span.end_time)
+
+            attributes = _extract_attributes(span.attributes)
+
+            cloud_trace_spans.append(
+                {
+                    "name": span_name,
+                    "span_id": span_id,
+                    "display_name": _get_truncatable_str_object(span.name),
+                    "start_time": start_time,
+                    "end_time": end_time,
+                    "parent_span_id": parent_id,
+                    "attributes": attributes,
+                    "links": _extract_links(span.links),
+                    "status": _extract_status(span.status),
+                    "time_events": _extract_events(span.events),
+                }
+            )
+            # TODO: Leverage more of the Cloud Trace API, e.g.
+            #  same_process_as_parent_span and child_span_count
+
+        return cloud_trace_spans
+
+    def shutdown(self):
+        pass
+
+
+def _get_hexidecimal_trace_id(trace_id: int) -> str:
+    return "{:032x}".format(trace_id)
+
+
+def _get_hexidecimal_span_id(span_id: int) -> str:
+    return "{:016x}".format(span_id)
+
+
+def _get_time_from_ns(nanoseconds: int) -> Dict:
+    """Given epoch nanoseconds, split into epoch milliseconds and remaining
+    nanoseconds"""
+    if not nanoseconds:
+        return None
+    seconds, nanos = divmod(nanoseconds, 1e9)
+    return {"seconds": int(seconds), "nanos": int(nanos)}
+
+
+def _get_truncatable_str_object(
+    str_to_convert: str, max_length: int = MAX_LENGTH
+):
+    """Truncate the string if it exceeds the length limit and record the
+    truncated bytes count."""
+    truncated, truncated_byte_count = _truncate_str(str_to_convert, max_length)
+
+    return TruncatableString(
+        value=truncated, truncated_byte_count=truncated_byte_count
+    )
+
+
+def _truncate_str(
+    str_to_check: str, limit: int = MAX_LENGTH
+) -> Tuple[str, int]:
+    """Check the length of a string. If exceeds limit, then truncate it."""
+    str_bytes = str_to_check.encode("utf-8")
+    str_len = len(str_bytes)
+    truncated_byte_count = 0
+
+    if str_len > limit:
+        truncated_byte_count = str_len - limit
+        str_bytes = str_bytes[:limit]
```

Comment:
It looks to me like this truncates the string at a byte boundary rather than a UTF-8 character boundary.
The decode() invocation below may fix that, but you may now end up with truncated_byte_count being a byte or two off. Not a blocker for thus PR, but would be good to test, unless there is a well-tested utility function for this somewhere that we could use instead.

Reply:
this is a really good point. I don't know a lot about cloud trace, but any reason to not truncate the string and then encode?